plasma_framework/contracts/src/transactions/PaymentTransactionModel.sol

```diff
@@ -41,36 +39,47 @@ library PaymentTransactionModel {
      * @return A decoded PaymentTransaction struct
      */
     function decode(bytes memory _tx) internal pure returns (PaymentTransactionModel.Transaction memory) {
-        RLPReader.RLPItem[] memory rlpTx = _tx.toRlpItem().toList();
-        require(rlpTx.length == ENCODED_LENGTH, "Invalid encoding of transaction");
-
-        RLPReader.RLPItem[] memory rlpInputs = rlpTx[1].toList();
-        require(rlpInputs.length <= MAX_INPUT_NUM, "Transaction inputs num exceeds limit");
-
-        RLPReader.RLPItem[] memory rlpOutputs = rlpTx[2].toList();
-        require(rlpOutputs.length <= MAX_OUTPUT_NUM, "Transaction outputs num exceeds limit");
+        GenericTransaction.Transaction memory btx = GenericTransaction.decode(_tx);
+        return fromGeneric(btx);
+    }
 
-        uint txType = rlpTx[0].toUint();
-        require(txType > 0, "Transaction type must not be 0");
+    function fromGeneric(GenericTransaction.Transaction memory btx) internal pure returns (PaymentTransactionModel.Transaction memory) {
+        require(btx.inputs.length <= MAX_INPUT_NUM, "Transaction inputs num exceeds limit");
+        require(btx.outputs.length <= MAX_OUTPUT_NUM, "Transaction outputs num exceeds limit");
 
-        bytes32[] memory inputs = new bytes32[](rlpInputs.length);
-        for (uint i = 0; i < rlpInputs.length; i++) {
-            bytes32 input = rlpInputs[i].toBytes32();
+        bytes32[] memory inputs = new bytes32[](btx.inputs.length);
+        for (uint i = 0; i < btx.inputs.length; i++) {
+            bytes32 input = btx.inputs[i].toBytes32();
             // Disallow null inputs
             require(uint256(input) != 0, "Null input not allowed");
             inputs[i] = input;
         }
 
-        PaymentOutputModel.Output[] memory outputs = new PaymentOutputModel.Output[](rlpOutputs.length);
-        for (uint i = 0; i < rlpOutputs.length; i++) {
-            PaymentOutputModel.Output memory output = PaymentOutputModel.decode(rlpOutputs[i]);
-            require(output.outputType != 0, "Output type must not be 0");
-            require(output.amount != 0, "Output amount must not be 0");
-            outputs[i] = output;
+        GenericTransaction.Output[] memory outputs = new GenericTransaction.Output[](btx.outputs.length);
+        for (uint i = 0; i < btx.outputs.length; i++) {
+            outputs[i] = decodeOutput(btx.outputs[i]);
         }
 
-        bytes32 metaData = rlpTx[3].toBytes32();
+        bytes32 metaData = btx.txData.toBytes32();
 
-        return Transaction({txType: txType, inputs: inputs, outputs: outputs, metaData: metaData});
+        return Transaction({txType: btx.txType, inputs: inputs, outputs: outputs, metaData: metaData});
+    }
+
+    function decodeOutput(RLPReader.RLPItem memory output) internal pure returns (GenericTransaction.Output memory) {
+        RLPReader.RLPItem[] memory outputRlpList = output.toList();
+        require(outputRlpList.length == 4, "Output must have 4 items");
```

Comment:
The difference is that a PaymentTransaction.Output must have exactly 4 items, whereas a GenericTransaction.Ouput may have more than 4 (if the concrete output format extends the struct and adds another field).

Maybe a better way to do this would be to follow the same model as the Transaction struct, and have an `outputData` field that can be extended as necessary?
However, if we do add this, it can't be optional - PaymentTransactions would have to always include a value (e.g. `0`) for this field.

What do you think?

Reply:
Ah...I see the difference now.

Btw, Do you mean the following structure?

```
struct Output {
        uint256 outputType;
        bytes20 outputGuard;
        address token;
        uint256 amount;
        RLPReader.RLPItem outputData;
}
```
Yeah....I am a bit not sure whether it justify the non-optional part. Though ugly, currently it is possible to use `GenericTransaction.decodeOutput` to get the first four fields first and then decode further if the output holds more field.

However, after seeing your trick of using RlpItem in transaction, If it is even possible I would be reeeeeally happy if the following is possible without bringing too much mess:

```
struct Output {
        uint256 outputType;
        RLPReader.RLPItem outputData;
}
```
And then introduce sth like `FungibleTokenOutputModel` to cover the `outputData`:
```
function getOutput(GenericTransaction.Transaction memory transaction, uint16 outputIndex) internal pure returns (FungibleOutput memory);

function decodeOutput(RLPItem memory outputRlp) internal pure returns (FungibleOutput memory);
```

Then our assumptions in implementation would make a lot more sense that `PaymentTx` only deal with fungible token outputs. This would make supporting ERC721 a lot easier with this structure, just need to introduce another `NonFungibleTokenOutputModel`

Reply:
Yes, I like this idea too 😄 
I don't think it's a huge change, and now is definitely the time to do it. Let me try it out and see what it looks like

Reply:
@thec00n 

> What's the rationale of introducing FungibleTokenOutputModel as part of this PR?

Here is the discussion of the model design👆

Reply:
Previous design of `WireTransaction` has the issue of not being really generalized enough as all outputs need to conform to same data structure. We've already seen that it would be not easy to support multi ERC721 <> ERC20 atomic swap tx.

We did not realize this RLPItem field trick until now. By leaving the responsibility to each output model itself, it leave us the ability to add whatever new output model we need in the future.